### PR TITLE
LDAP Wizard Overhaul

### PR DESCRIPTION
**Pull request by blizzz**
This is the next generation wizard.

It contains:
- a big refactor of the mostly monolithic old monster
- raw filter input fields are now text areas
- diffrerent selection elements for directories with many groups
- port and base dn detection triggered only by button click
- base DN test
- Add configuration was removed from the configuration chooser and instead split into "new" and "clone" buttons next to it
- wording changes
- more helpful information in labels and tooltips
- limits for LDAP side searches
- LDAP filter also visibile in assisted mode
- no automatic user and group count, to be triggered by button click
- test for the login filter

Also (possibly) fixes:
- https://github.com/owncloud/core/issues/13904
- https://github.com/owncloud/core/issues/11504
- https://github.com/owncloud/core/issues/15444
- https://github.com/owncloud/core/issues/11571
- https://github.com/owncloud/core/issues/11447
- https://github.com/owncloud/core/issues/11445
- https://github.com/owncloud/core/issues/11330
- https://github.com/owncloud/core/issues/7908
- https://github.com/owncloud/core/issues/14244

@karlitschek @DeepDiver1975

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
@blizzz unit test on ldap fail - FYI

**Comment by DeepDiver1975** [comment]
https://ci.owncloud.org//job/pull-request-analyser-ng-simple/11334/

**Comment by blizzz** [comment]
forgot to adjust the tests :-/ but here we go.

**Comment by blizzz** [comment]
Please review @PVince81 @MorrisJobke @Xenopathic @jnfrmarks 

@jnfrmarks we should talk about testing it

cc @MTRichards @craigpg

File: apps/user_ldap/img/copy.svg [hunk 1]
```diff
@@ -0,0 +1,120 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (MorrisJobke) [inline apps/user_ldap/img/copy.svg]: @jancborchardt you have such nice optimisation scripts ;)

**Comment by karlitschek** [comment]
:+1: @craigpg @MTRichards What do you think?

File: apps/user_ldap/img/copy.svg (hunk 1 above)
Comment (blizzz) [inline apps/user_ldap/img/copy.svg]: Ideally someone conjures an icon of our own for copy/clone . I took this now as being short of time.

File: apps/user_ldap/js/wizard/wizardDetectorTestLoginName.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * This file is licensed under the Affero General Public License version 3 or later.
+ * See the COPYING-README file.
+ */
+
+OCA = OCA || {};
+
+(function() {
+
+	/**
+	 * @classdesc a Port Detector. It executes the auto-detection of the port
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardDetectorTestLoginName.js]: Copy and paste error?

File: apps/user_ldap/js/wizard/wizardDetectorQueue.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+			var detectorQueue = this;
+			$.when(request).then(function() {
+				detectorQueue.isRunning = false;
+				detectorQueue.next();
+			});
+		}
+	});
+
+	OCA.LDAP.Wizard.WizardDetectorQueue = WizardDetectorQueue;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardDetectorQueue.js]: Newline

File: apps/user_ldap/js/wizard/wizardFilterOnType.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+				if(!filterVal || this.normalized.indexOf(filterVal) > -1) {
+					fity.$select.append($('<option>').val(this.value).text(this.value));
+				}
+			});
+			delete(fity.runID);
+		}
+	});
+
+	OCA.LDAP.Wizard.FilterOnType = FilterOnType;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardFilterOnType.js]: newline

File: apps/user_ldap/js/wizard/wizardFilterOnTypeFactory.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+		 * @param {jQuery} $select
+		 * @param {jQuery} $textInput
+		 */
+		get: function($select, $textInput) {
+			return new OCA.LDAP.Wizard.FilterOnType($select, $textInput);
+		}
+	});
+
+	OCA.LDAP.Wizard.FilterOnTypeFactory = FilterOnTypeFactory;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardFilterOnTypeFactory.js]: Newline

File: apps/user_ldap/js/wizard/wizardTabAbstractFilter.js [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+			this._saveGroups(selected);
+			$selected.find('option:selected').appendTo($available);
+			this.updateFilterOnType();
+		}
+
+	});
+
+	OCA.LDAP.Wizard.WizardTabAbstractFilter = WizardTabAbstractFilter;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabAbstractFilter.js]: newline

File: apps/user_ldap/js/wizard/wizardTabAdvanced.js [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+		 */
+		onResultReceived: function(view, payload) {
+			if(payload.feature === 'TestConfiguration') {
+				OC.Notification.showTemporary(payload.data.message);
+			}
+		}
+	});
+
+	OCA.LDAP.Wizard.WizardTabAdvanced = WizardTabAdvanced;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabAdvanced.js]: newline

File: apps/user_ldap/js/wizard/wizardTabElementary.js [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+			$('#ldap_action_copy_configuration').click(function(event) {
+				event.preventDefault();
+				view.configModel.newConfig(true);
+			});
+		}
+	});
+
+	OCA.LDAP.Wizard.WizardTabElementary = WizardTabElementary;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabElementary.js]: newline

File: apps/user_ldap/js/wizard/wizardTabExpert.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+				} else {
+					message = t('user_ldap', 'Error while clearing the mappings.');
+				}
+				OC.Notification.showTemporary(message);
+			}
+		}
+	});
+
+	OCA.LDAP.Wizard.WizardTabExpert = WizardTabExpert;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabExpert.js]: newline

File: apps/user_ldap/js/wizard/wizardTabGroupFilter.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+		 * @returns {string}
+		 */
+		getFilterModeKey: function () {
+			return 'ldap_group_filter_mode';
+		}
+
+	});
+
+	OCA.LDAP.Wizard.WizardTabGroupFilter = WizardTabGroupFilter;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabGroupFilter.js]: newline

File: apps/user_ldap/js/wizard/wizardTabLoginFilter.js [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+				OC.Notification.showTemporary(t('user_ldap', 'Please provide a login name to test against'), 3);
+			} else {
+				this.configModel.requestWizard('ldap_test_loginname', {ldap_test_loginname: testLogin});
+			}
+		}
+
+	});
+
+	OCA.LDAP.Wizard.WizardTabLoginFilter = WizardTabLoginFilter;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabLoginFilter.js]: newline

File: apps/user_ldap/js/wizard/wizardTabUserFilter.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+			) {
+				message = t('user_ldap', 'The group box was disabled, because the LDAP / AD server does not support memberOf.');
+			}
+			return message;
+		}
+
+	});
+
+	OCA.LDAP.Wizard.WizardTabUserFilter = WizardTabUserFilter;
+})();
```
Comment (MorrisJobke) [inline apps/user_ldap/js/wizard/wizardTabUserFilter.js]: newline

File: apps/user_ldap/lib/wizard.php [hunk 13]
```diff
@@ -467,8 +460,9 @@ public function determineGroupMemberAssoc() {
 			return false;
 		}
 		$this->configuration->setConfiguration(array('ldapGroupMemberAssocAttr' => $attribute));
+		$this->result->addChange('ldap_group_member_assoc_attribute', $attribute);
 		//so it will be saved on destruct
-		$this->result->markChange();
+		//$this->result->markChange();
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/wizard.php]: ?

File: apps/user_ldap/lib/wizard.php [hunk 14]
```diff
@@ -720,7 +752,7 @@ private function checkHost() {
 	 * @throws \Exception
 	 */
 	private function detectGroupMemberAssoc() {
-		$possibleAttrs = array('uniqueMember', 'memberUid', 'member', 'unfugasdfasdfdfa');
+		$possibleAttrs = array('uniqueMember', 'memberUid', 'member');
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/wizard.php]: :laughing:

File: apps/user_ldap/settings.php [hunk 15]
```diff
@@ -29,13 +29,52 @@
...
 OC_Util::checkAdminUser();
 
-OCP\Util::addScript('user_ldap', 'ldapFilter');
-OCP\Util::addScript('user_ldap', 'experiencedAdmin');
-OCP\Util::addScript('user_ldap', 'settings');
 \OC_Util::addVendorScript('user_ldap', 'ui-multiselect/src/jquery.multiselect');
-OCP\Util::addStyle('user_ldap', 'settings');
 \OC_Util::addVendorStyle('user_ldap', 'ui-multiselect/jquery.multiselect');
 
+\OCP\Util::addScript('user_ldap', 'wizard/controller');
```
Comment (MorrisJobke) [inline apps/user_ldap/settings.php]: Can you do this all inside the template? Would be really nice

`script()`, `vendor_script()`, `style()`, `vendor_style()`

Comment (MorrisJobke) [inline apps/user_ldap/settings.php]: And they accept an array as second parameter - then you don't need to repeate `user_ldap` ;)

File: apps/user_ldap/settings.php [hunk 16]
```diff
@@ -81,4 +120,4 @@
 	$tmpl->assign($key.'_default', $default);
 }
 
-return $tmpl->fetchPage();
+return $tmpl->fetchPage();
```
Comment (MorrisJobke) [inline apps/user_ldap/settings.php]: New line ;)

File: apps/user_ldap/templates/wizard-container.php [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<form id="ldapWizard" class="section" action="#" method="post">
+	<h2><?php p($l->t('LDAP2')); ?></h2>
+</form>
```
Comment (MorrisJobke) [inline apps/user_ldap/templates/wizard-container.php]: new line

**Comment by MorrisJobke** [comment]
That one is huge! Let me test it :)

File: apps/user_ldap/js/wizard/wizardDetectorTestLoginName.js (hunk 2 above)
Comment (blizzz) [inline apps/user_ldap/js/wizard/wizardDetectorTestLoginName.js]: yup, used the other as template

**Comment by blizzz** [comment]
@MorrisJobke your concerns are addressed.

**Comment by blizzz** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
* [ ] clicking the tab headers doesn't update the back, next buttons (clicking the first when in the group filter results in only the back button, which is useless for the first tab)

**Comment by MorrisJobke** [comment]
Looks good so far. Seems to be more robust that the previos one.

**Comment by LukasReschke** [comment]
- [ ] When creating a new server configuration the previously selected config is shown twice. Well or it does some other confusing stuff as can be seen in my screenshot where it shows the "1" at the top and at the end.



To reproduce:

1. Go to the admin settings
2. Click the plus a few times
3. Click on the drop down

File: apps/user_ldap/js/wizard/wizardTabLoginFilter.js [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+		 * @param result
+		 */
+		handleLoginTestResult: function(result) {
+			var message;
+			var isHtml = false;
+			if(result.status === 'success') {
+				var usersFound = parseInt(result.changes.ldap_test_loginname, 10);
+				if(usersFound < 1) {
+					var filter = result.changes.ldap_test_effective_filter;
+					message = t('user_ldap', 'User not found. Please check your login attributes and username. Effective filter (to copy-and-paste for command line validation): <br/>' + filter);
```
Comment (LukasReschke) [inline apps/user_ldap/js/wizard/wizardTabLoginFilter.js]: While very likely not exploitable at all in the real world: Would it make sense to escape `filter`?

Comment (blizzz) [inline apps/user_ldap/js/wizard/wizardTabLoginFilter.js]: absolutely – addressed!

**Comment by blizzz** [comment]
@LukasReschke wrt https://github.com/owncloud/core/pull/15492#issuecomment-91214206: i could not reproduce any duplications, but wrong insert position and number. This was addressed. Could you check whether this fixes what you see?

**Comment by blizzz** [comment]
@MorrisJobke wrt https://github.com/owncloud/core/pull/15492#issuecomment-91212462: fixed!

**Comment by blizzz** [comment]
> Looks good so far. Seems to be more robust that the previos one.

That was my big goal :dancer:

**Comment by blizzz** [comment]
@owncloud-bot retest this please

**Comment by blizzz** [comment]
No idea what's up with Jenkins but there are no failing tests.

**Comment by MorrisJobke** [comment]
> No idea what's up with Jenkins but there are no failing tests.

```
09:09:45  > git config core.sparsecheckout # timeout=10
09:09:45  > git checkout -f 7996e8ad9fe52e7c067bc6538a89da6d779da4cb
09:09:45 FATAL: Could not checkout null with start point 7996e8ad9fe52e7c067bc6538a89da6d779da4cb
09:09:45 hudson.plugins.git.GitException: Could not checkout null with start point 7996e8ad9fe52e7c067bc6538a89da6d779da4cb
09:09:45 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$8.execute(CliGitAPIImpl.java:1855)
09:09:45 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:152)
09:09:45 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:145)
09:09:45 	at hudson.remoting.UserRequest.perform(UserRequest.java:121)
09:09:45 	at hudson.remoting.UserRequest.perform(UserRequest.java:49)
09:09:45 	at hudson.remoting.Request$2.run(Request.java:324)
```

Checkout failure ... maybe you need a jenkins branch or simply retest harder ;)

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by LukasReschke** [comment]
> @LukasReschke wrt #15492 (comment): i could not reproduce any duplications, but wrong insert position and number. This was addressed. Could you check whether this fixes what you see?

Acknowledged. Works fine now.

**Comment by blizzz** [comment]
Test passed in the Jenkins-dedicated branch

**Comment by mmattel** [comment]
The hint text shown in field "Einen Basis-DN...." refers to a non existing tab "Erweitert"

image

**Comment by blizzz** [comment]
@mmattel this is a translation bug. "Erweitert" should be "Fortgeschritten" (or vice versa). Please check over at https://www.transifex.com/projects/p/owncloud/

**Comment by mmattel** [comment]
I have changed in https://www.transifex.com/projects/p/owncloud/viewstrings/#de/user_ldap/12396160
from ``Erweitert`` to ``Fortgeschritten``

**Comment by blizzz** [comment]
@mmattel cool, thank you :)

**Comment by MorrisJobke** [comment]
* [ ] clicking through via next the login filter isn't set initially - I need to uncheck and then check again the checkbox "LDAP / AD Username:"
* [ ] Having an error in there the state doesn't change to green if the error is fixed - reload of page is required to have a green circle (without changing anything)

**Comment by MorrisJobke** [comment]
Beside the above issues this is ready from my side :)

**Comment by MorrisJobke** [comment]
Once they are done I will test again.

**Comment by blizzz** [comment]
@MorrisJobke issue one is solved, the second one not, because I could not reproduce it yet.

**Comment by blizzz** [comment]
@MorrisJobke maybe 2nd was a side effect of another issue i found and fixed. Please test again.

**Comment by blizzz** [comment]
Tests passed: https://github.com/owncloud/core/pull/15516#issuecomment-91335075

**Comment by blizzz** [comment]
Nothing urgent or bad in scrutinizers findings.

Needs one reviewer/thumb to get merged.

**Comment by MorrisJobke** [comment]
Both problems are gone. Works a lot more smooth now :) :+1:

**Comment by blizzz** [comment]
Thx @MorrisJobke

**Comment by karlitschek** [comment]
awesome @blizzz :-)

**Comment by blizzz** [comment]
FYI, i did a test plan: https://s3.owncloud.com/owncloud/public.php?service=files&t=qxJIlxZtXRmYD8A

**Pull request merged**